[FEAT] 캘린더 공통 컴포넌트 추가

geongyu09 opened the pull request:
## 주요 변경사항 ✔️

홈 화면의 캘린더 부분을 개발하다가 너무 변경사항이 커지는 것 같아서 캘린더 공통 컴포넌트 추가까지 하여 pr을 우선적으로 날립니다. 

주된 변경사항은 아래와 같습니다. 
- 일정 가져오는 api 추가 및 목서버 로직 추가
- full calendar를 사용한 Calendar 공통 컴포넌트 추가
  - (*추가만 하였습니다. 기능적으로는 read 부분만 일부 완성되었습니다.)

### /service 폴더
우선 외부 라이브러리에 의존성을 가지는 로직은 따로 분리하고, 이를 사용하는 곳에서는 인터페이스를 통해서 사용할 수 있도록 하는 것이 중요하다고 생각하였습니다. 
여기서는 full calendar를 사용하여서 캘린더 ui를 구현하는데, 만약 full-calendar가 아니라 toast와 같은 다른 라이브러리로 변경하거나, 혹은 full-calendar 에 문제가 생겼을 때, 기존의 우리의 로직을 살펴보거나 수정할 필요가 없어진다는 장점을 기대할 수 있습니다. 
이를 어떻게 다른 로직으로 분리하는 것이 좋을까에 대해서 고민을 하다가 2단계에서 사용하지 않았던 service 폴더를 새롭게 만들기로 하였습니다. 
(기존 다른 폴더들을 고려해보았지만, 이 로직이 들어갈만한 적당한 곳을 찾을 수 없었습니다.)

제가 생각한 service 폴더의 쓰임은 아래와 같습니다. 
> /service : 외부 라이브러리와 같은 서비스 로직을 따로 다루는 디랙토리

### Calendar 
캘린더의 경우 여러 곳에서 재사용할 수 있도록 만들어야 했습니다. 하지만 어디에서 어떻게 사용하는지에 대해서 올바르게 알 수가 없었고, 우선적으로 현재 구현중인 홈 화면에 맞추어 구현하지만, 확장에 열려있을 수 있도록 고안하고 싶었습니다. 

**[폴더 구조]**
우선적으로 아래의 폴더구조를 가지고 있습니다. 

```
- components: 컴포넌트들이 담긴다.
- style : css 파일들이 담긴다.
- types : 타입들이 담긴다.
- utils : 유틸 파일들이 담긴다.
- index.ts : 모듈화 되어있는 파일들을 재모듈화 한다.
```
만약 성격이 많이 다른 컴포넌트를 만든다면 새로운 컴포넌트를 components에 만들고, 이를 index.ts에 재 수출하는 방식으로 구현할 수 있을 것이라 생각하였습니다. 

## 리뷰어에게... 🙏

- 해당 이슈가 전부 해결된 것은 아닙니다. 이후에 해당 pr에서 다시 작업을 하거나, 혹은 새로운 pr을 파서 이후의 작업을 하도록 하겠습니다. 
현재 작성된 부분은 CRUD중 R의 일부만 작성된 것입니다. 이점 유의해주시면 감사하겠습니다

### 추가 변경사항

추가적으로 스타일링을 추가 및 피드백 받은 부분을 수정하였습니다. 
주된 변경사항은 아래와 같습니다. 
- 기본적으로 사용하는 style var 값을 variants.css 에 정의하였습니다. 
  - 여기서 정의한 값은 컴포넌트에서 (특히나 emotion에서) 사용하기 쉽도록 js 변수로서 가져와 export 하였습니다. (variants.ts)
- 캘린더의 기본으로 정의되어있던 css 변수값을 수정하여 스타일링을 추가하였습니다. 

crud가 전부 있는 컴포넌트였지만, 저번 회의를 통해서 R만 하게 되어 별다른 추가적인 작업은 없었습니다! 
간단히 디자인을 해보았는데, 내부 구현사항을 변경하는데 비용이 많이 들것으로 예상하여 기본적으로 제공하는 UI 내에서 수정하는 방식으로 하게 되었습니다.
그러다보니 기존의 피그마 와이어프레임과 다른 점이 존재하는데, 양해보탁드립니다.. 
디자인 확인해보시고 수정할 점 있다면 말씀해주세요!
추가적인 변경사항까지 확인 되면 머지하도록 하겠습니다

![image](https://github.com/user-attachments/assets/6817d1b7-d049-4297-9e60-d122e8cf840e)

![image](https://github.com/user-attachments/assets/830c34b0-9eb9-408d-b4dc-85bba0f631af)


## 관련 이슈
#35

geongyu09 committed:
chore(package.json): 캘린더 구현을 위한 full-calendar 의존성 설치

#35

geongyu09 committed:
feat(useGetMyEvent): 홈 화면에서 사용하는 내 캘린더 가져오기 로직 추가

#35

geongyu09 committed:
feat(calendar.mock.ts): 캘린더 가져오기 목서버 로직 추가

geongyu09 committed:
feat(package.json): full-calendar 의존성 추가

#35

geongyu09 committed:
fix: 올바르게 axios타입이 지정되지 않았던 버그 수정

기존에 resopnse에 존재하던 탬플릿 타입이 기록이 안되어 ts타입과 실제 타입이 맞지 않던 문제 발생

instance에서 이를 해결하기 위하여 data만 사용하도록 수정

#35

geongyu09 committed:
feat(calendar): 반복되어 사용될, 외부 라이브러리 의존성을 가지는 캘린더 UI 컴포넌트 추가

외부 라이브러리에 의존하므로 이를 Service폴더로 뺌
구조의 경우 아래와 동일하다.
- components: 컴포넌트들이 담긴다
- style : css 파일들이 담긴다.
- types : 타입들이 담긴다
- utils : 유틸 파일들이 담긴다
- index.ts : 모듈화 되어있는 파일들을 재모듈화 한다.

현재 utils 내용의 경우 추후 추가하도록 한다.

geongyu09 committed:
feat(Home/index.tsx): 홈 화면에 캘린더 추가

#35

Da-Yeon-Kim commented on src/api/hooks/useGetMyEvents.ts:
```diff
@@ -0,0 +1,42 @@
+
+export const getMyEvent = async () => {
+  const response = await fetchWithToken<Event[]>(getMyEventPath());
+  return response.data;
```
instace 파일에서 이미 `response.data`로 받아오도록 수정하셨는데, hooks 폴더 내의 파일(`useGetMyEvents`, `useGetUser` 등)에서도 `.data`를 이중으로 붙여서 사용하고 있네요. 확인 한 번 해주실래요?

geongyu09 replied:
instance에서 data를 벗긴건 axios 반환 객체에서 data를 벗긴 부분입니다. 
저희가 받아오는 response는 data 필드로 한 뎁스 들어가 있기에 이중으로 벗기는 것이 맞습니다!

Da-Yeon-Kim commented on src/service/Calendar/utils/converter.ts:
```diff
@@ -0,0 +1 @@
+export default function Converter() {}
```
아직 작성을 안 하신 거죠...? 혹시 추후에 무슨 역할 하는 코드인지 알려주실 수 있나요? Converter니까 데이터 형식 변환... 할 때 쓰는 코드인가요?

geongyu09 replied:
네 추후에 작성할 부분입니다! 
커밋에는 설명한 부분이지만 해당 부분을 pr 에 남기지 않았었네요

Da-Yeon-Kim commented on src/service/Calendar/components/DefaultCalendar/index.tsx:
```diff
@@ -0,0 +1,64 @@
+
+type Props = {
+  height?: number;
+  evnets?: Event[];
```
ㅎㅎ 여기 오타가 있네요! `evnets`로 오타 났어요~ 엇 근데 밑에서도 `evnets`로 사용하고 계시는데,, 혹시 의도적인 작명인가요?!

geongyu09 replied:
변경하겠습니다 감사합니다!
아래 부분은 ts를 사용하니까 추론해서 사용하다보니 그렇게 된 것 같습니다

Da-Yeon-Kim commented on src/components/features/Home/MyCalendar/index.tsx:
```diff
@@ -0,0 +1,28 @@
+    return <div>Error</div>;
+  }
+
+  if (status === 'pending') {
```
혹시 `pending`에 대해 설명해주실 수 있나요...? `loading`을 주로 사용했어서,, 찾아보니 보류 중인?? 처리 중인?? 같은 의미인 것 같은데 `react-query`에서도 사용하는 상태값인가요?

geongyu09 replied:
기존에 loading상태는 !data 를 보장해주지 않습니다.
로딩이 끝나면 data는 undefined가 아닌 실존하는 값일텐데 말이죠. 
그렇기 때문에 기존대로 로직을 짜게 되면 if (isLoading || !data) 로 작성해야 하고, 이는 직관적이지만, 흔히 생각하는 저희의 상식과 맞지는 않습니다. 

이를 tanstack 팀에서도 인지하였고, v5에서 이 두가지를 합해서 pending상태를 만들었습니다.
공식문서 보시면 왜 변경하게 됐고, 이를 왜 사용해야하는지에 대해서 저보다 잘 설명할 테니 공식문서를 참고해주세요!


https://tanstack.com/query/latest/docs/framework/react/guides/migrating-to-v5#status-loading-has-been-changed-to-status-pending-and-isloading-has-been-changed-to-ispending-and-isinitialloading-has-now-been-renamed-to-isloading

Da-Yeon-Kim replied:
덕분에 좋은 지식 하나 얻고 가네요ㅎㅎ 친절한 답변 감사해요!

Da-Yeon-Kim commented on src/components/features/Home/MyCalendar/index.tsx:
```diff
@@ -0,0 +1,28 @@
+    editable: true,
+  }));
+
+  return <DefaultCalendar evnets={events} />;
```
여기도 `evnets`이 있는데 진짜 의도한 작명인가요...?!

geongyu09 replied:
수정하겠습니다!

Da-Yeon-Kim commented on src/components/features/Home/MyCalendar/index.tsx:
```diff
@@ -0,0 +1,28 @@
+    return <div>Loading...</div>;
+  }
+
+  const events: Event[] = data.map(({ id, title, color, time: { start_at, all_day, end_at } }) => ({
```
데이터가 없을 경우를 대비해서 `data?.map`하여 데이터 없으면 빈 배열로 처리해주면 좋을 것 같아요~

geongyu09 replied:
이전 답글과 동일한 내용인데요, pending을 사용했기 때문에 데이터는 없는 경우는 없습니다!
서버에서 반환하는 데이터가 명세서를 따르지 않는 경우를 제외하면 타입을 지정한대로 받을 수 있어요!

추가적으로 해당 부분이 undefined와 같은 값이라면 lint에서 해당 부분을 잡아줍니다. 
그렇게 되면 빨간줄이 뜰거고(정적 테스트) 당연히 빌드는 안될꺼에요(npm start 안됨)

이 부분은 리뷰하기전에 돌려보셔도 좋을 것 같아요..!

Da-Yeon-Kim reviewed (commented):
외부 라이브러리 로직을 `/service` 폴더로 따로 분리하는 방식 좋은 것 같습니다, 음음!
캘린더가 워낙 구현할 기능도 많고, 외부 라이브러리도 써야 해서 머리 아플 텐데 고생이 많으셔요~ 파이팅💪

+) Weekly5의 merge 정책이 적용 안 된 것 같아요! 이거 참 브랜치 새로 팔 때마다 매번 적용하는 게 꽤 번거롭네요ㅎㅎ

HyeonChae1104 reviewed (approved):
service 폴더 생성해 작업하신다는 내용 확인하였습니다! 
폴더 구조와 쓰임도 정확히 적어주셔서 더 좋네요! 수고많으셨습니다!!

Da-Yeon-Kim reviewed (approved):
추후에 작업하실 때 오타만 고치시면 될 것 같네요~~ 고생하셨어요! 앗차 머지 정책도요!

geongyu09 committed:
feat(Calendar): 요구사항의 변경에 따라 불필요한 코드 제거

수정, 삭제가 제거됨에 따라서 불필요한 로직 제거

geongyu09 committed:
refactor(Calendar/utils): 사용하지 않는 유틸 코드 제거

#35

geongyu09 committed:
design(Calendar): 캘린더 디자인 추가

- 자주 사용되는 서비스의 primary 색 변수를 추가하였습니다.
  - 색 변수는 css로 정의되어있으며, 이는 styles/variants/index.ts에서 js로 변환 후 가져오는 구조를 가집니다.
- 캘린더의 기본 스타일을 수정하였습니다.

#35

geongyu09 committed:
refactor: event 오타 수정

geongyu09 commented:
(위에도 적어두었지만, 가독성을 위해 아래에 다시 첨부합니다. 해당 내용 확인해주시고 다시 approve 해주세요!)
### 추가 변경사항

추가적으로 스타일링을 추가 및 피드백 받은 부분을 수정하였습니다. 
주된 변경사항은 아래와 같습니다. 
- 기본적으로 사용하는 style var 값을 variants.css 에 정의하였습니다. 
  - 여기서 정의한 값은 컴포넌트에서 (특히나 emotion에서) 사용하기 쉽도록 js 변수로서 가져와 export 하였습니다. (variants.ts)
- 캘린더의 기본으로 정의되어있던 css 변수값을 수정하여 스타일링을 추가하였습니다. 

crud가 전부 있는 컴포넌트였지만, 저번 회의를 통해서 R만 하게 되어 별다른 추가적인 작업은 없었습니다! 
간단히 디자인을 해보았는데, 내부 구현사항을 변경하는데 비용이 많이 들것으로 예상하여 기본적으로 제공하는 UI 내에서 수정하는 방식으로 하게 되었습니다.
그러다보니 기존의 피그마 와이어프레임과 다른 점이 존재하는데, 양해보탁드립니다.. 
디자인 확인해보시고 수정할 점 있다면 말씀해주세요!
추가적인 변경사항까지 확인 되면 머지하도록 하겠습니다

![image](https://github.com/user-attachments/assets/6817d1b7-d049-4297-9e60-d122e8cf840e)

![image](https://github.com/user-attachments/assets/830c34b0-9eb9-408d-b4dc-85bba0f631af)

Da-Yeon-Kim commented on src/styles/variants/index.ts:
```diff
@@ -6,3 +8,8 @@ export const breakpoints = {
   xl: '1280px',
   '2xl': '1536px',
 };
+
```
아, 우와!
저도 모달 스타일링 하면서 테마 컬러를 CSS나 TS 다 쓸 수 있게 공통으로 관리 못 하나 생각했는데
이런 좋은 방법이 있네요! 완전 👍
이거 머지되고 나면 저도 쇽샥 가져다 써야겠어요ㅎㅎ

그 와중에 컬러 네이밍 센스가 서로 다른 게 웃기네요ㅋㅋㅋ
(그냥 green이라고 지은 사람)

Da-Yeon-Kim reviewed (approved):
저도 모달 스타일링 하면서 생각했는데,
아~ 이거 외부 라이브러리 스타일링 하는 게 쉬운 일이 아니더라고요ㅋㅋㅋ
공식 문서 뚫어져라 쳐다 보면서 어떻게 바꿀지 고민하는 게... 진짜 고생하셨어요!

Figma 그대로 옮기기보단 코드 짜면서 적당히 바꾸는 게 옳은 것 같습니다.
비용 많이 드는 건 과감하게 out!
지금 이 스타일도 충분히 메인 테마와 통일성 있고, 기능 구현에도 모자람 없어서 좋은 것 같아요~ 👍

geongyu09 committed:
fix: npm 버전이 달라서 발생한 package-lock.json 파일 제거

추후에 이슈 파서 버전을 맞추도록 하겠습니다

#35

geongyu09 committed:
fix: 커밋 충돌 해결

geongyu09 committed:
feat: package-lock 파일 생성

geongyu09 committed:
충돌 해결
